ls1intum/Artemis
# Use Git Archive to Export Programming Exercise Repositories

### Checklist

### Motivation and Context


The fix implemented in PR #3257 doesn't fix the issue that occurs when exporting programming exercises during the course archival. For some reason, the garbage collection from JGit kicks in during the export of programming exercises which leads to the `java.nio.file.NoSuchFileException` being thrown on `gc.log.lock`. 

### Description

This PR fixes the issue by implementing a new way of exporting/archiving code repositories. It takes advantage of the native `git archive` command to generate a zip archive of the repository. Since we don't export repositories manually anymore and let JGit do the heavy-lifting, we hope that it will fix the issue mentioned above.

This PR also adds some small fixes:

1. It strips illegal characters from all the filenames used for the zip files in order to prevent problems with the OS
2. It adds more log messages during the archival process
3. It catches the `UncaughtIOException` which may occur during a JGit operation

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
4. Generate one (or more) exercise of each type
5. Participate in each exercise
6. Generate an exam with exercises of each type
7. Participate in the exam
8. Edit the course and change the end date so that it's in the past
9. Navigate to the course detail page and click on "archive course"
10. Once archiving is complete, you should be able to download the course archive
11. Download the archive. You should be able to see the submissions in the zip file


### Test Coverage



CourseExamExportService.java 85%
FileService.java 83%
SubmissionExportService.java 87%
GitService.java 75%
ProgrammingExerciseExportService.java 72%
